src/url-params.ts

```diff
@@ -0,0 +1,105 @@
+import {
+	getCookie,
+	hasCookieSupport,
+	setCookie,
+} from '@analytics/cookie-utils';
+import union from 'lodash/union';
+import { Mixpanel } from 'mixpanel-browser';
+
+const URL_PARAM_DEVICE_ID = 'd_id';
+const COOKIES_DEVICE_IDS = '__analytics_dids';
+
+const deviceIdSeparator = /\s*,\s*/;
+
+/**
+ * AnalyticsUrlParams helps with handling analytics-related URL parameters
+ * like anonymous device identifier.
+ */
+export class AnalyticsUrlParams {
+	private readonly cookiesSupported = hasCookieSupport();
+	private deviceIds: string[];
+
+	constructor(private mixpanel?: Mixpanel) {
+		const storedValue = this.cookiesSupported
+			? getCookie(COOKIES_DEVICE_IDS)
+			: null;
+		this.setDeviceIds(storedValue, null);
+	}
+
+	private setDeviceIds(
+		inputIdString: string | null,
+		currentDeviceId: string | null,
+	) {
+		const list = inputIdString ? inputIdString.split(deviceIdSeparator) : [];
+		if (currentDeviceId) {
+			list.push(currentDeviceId);
+		}
+		this.deviceIds = union(this.deviceIds, list);
+		if (this.cookiesSupported) {
+			setCookie(COOKIES_DEVICE_IDS, this.deviceIds.join(','));
+		}
+		return list.length > 0 ? list[0] : null;
+	}
+
+	clearCookies() {
+		if (this.cookiesSupported) {
+			setCookie(COOKIES_DEVICE_IDS, '');
+		}
+	}
+
+	/**
+	 * Analyzes the query string and stores the anonymous device ID if it's there.
+	 * Returns null if the current URL query string should not be modified after consuming the data.
+	 *
+	 * @param queryString URL query string to process
+	 * @return a modified query string that hides processed/consumed parameters if the query string should be modified
+	 */
+	consumeUrlParameters(queryString: string): string | null {
+		const params = new URLSearchParams(queryString);
+
+		const passedDeviceId = params.get(URL_PARAM_DEVICE_ID);
+		if (passedDeviceId) {
+			const originalMixpanelId =
+				this.mixpanel != null ? this.mixpanel.get_distinct_id() : null;
+
+			const newCurrentDeviceId = this.setDeviceIds(
```

Comment:
Normally you wouldn't expect a setter to return anything, and in this case returning the first element is a bit confusing. How about 
```
this.setDeviceIds(passedDeviceIds);
const newCurrentDeviceId = this.deviceIds ? this.deviceIds[0] : null;
```

If you set `this.deviceIds` to the mixpanel distinct ID in the constructor, `setDeviceIds` will only require a single parameter, as `mixpanelId` will already by in `deviceIds`.

Reply:
Please note that this is not a public setter, but a private method and an implementation detail.

If we remove parsing of the input string, we'll need to parse it separately in the constructor (because cookie value is a string) and in `consumeUrlParameters`.
Since it does the parsing, it also returns the first device ID that can be used to configure mixpanel.

That said, we can rename it so it's not confused with a setter semantics. To something like `updateDeviceIds`.

Reply:
I think whether a function is public or private doesn't matter much in terms of functions being intuitive in their behavior. In this case, you will just read from `this.deviceIds` after calling `setDeviceIds` instead of expecting a return parameter, that's all.

Reply:
>I think whether a function is public or private doesn't matter much in terms of functions being intuitive in their behavior

My main point is not about public/private, but about the fact that it's not necessarily a classic setter. 

Also, although `this.deviceId` is an array, it's really a set of device IDs: when setDeviceIds is called, we merge the existing list with a new one, removing duplicates. The order of elements is not really guaranteed, hence, we should not assume `this.deviceIds[0]` is the first ID from the URL query.